process/process.go

```diff
@@ -230,20 +235,61 @@ func (sp *systemProcess) GetMappings() ([]Mapping, uint32, error) {
 	defer mapsFile.Close()
 
 	mappings, numParseErrors, err := parseMappings(mapsFile)
-	if err == nil {
-		fileToMapping := make(map[string]*Mapping, len(mappings))
-		for idx := range mappings {
-			m := &mappings[idx]
-			if m.Inode == 0 {
-				// Ignore mappings that are invalid,
-				// non-existent or are special pseudo-files.
-				continue
-			}
-			fileToMapping[m.Path] = m
+	if err != nil {
+		return mappings, numParseErrors, err
+	}
+
+	if len(mappings) == 0 {
+		// Test for main thread exit by checking for Zombie state
+		pidStat, err := os.ReadFile(fmt.Sprintf("/proc/%d/stat", sp.pid))
+		if err != nil {
+			// Should never happen while process is alive
+			return nil, 0, err
+		}
+
+		var p int
+		var c string
+		var state rune
+		n, err := fmt.Sscanf(string(pidStat), "%d %s %c", &p, &c, &state)
+		if err != nil || n < 3 {
+			// Should never happen
+			return nil, 0, err
 		}
-		sp.fileToMapping = fileToMapping
+		if state != 'Z' {
+			return mappings, numParseErrors, ErrNoMappings
+		}
+
+		log.Warnf("PID: %v main thread exit", sp.pid)
+		sp.mainThreadExit = true
+
+		log.Warnf("TID: %v extracting mappings", sp.tid)
+		mapsFileAlt, err := os.Open(fmt.Sprintf("/proc/%d/task/%d/maps", sp.pid, sp.tid))
+		// On all errors resulting from trying to get mappings from a different thread,
+		// return ErrNoMappings which will keep the PID tracked in processmanager and
+		// allow for a future iteration to try extracting mappings from a different thread.
+		// This is done to deal with race conditions triggered by thread exits (we do not want
+		// the agent to unload process metadata when a thread exits but the process is still
+		// alive).
+		if err != nil {
+			return mappings, numParseErrors, ErrNoMappings
+		}
+		defer mapsFileAlt.Close()
+
+		numParseErrorsAlt := uint32(0)
+		mappings, numParseErrorsAlt, err = parseMappings(mapsFileAlt)
+		numParseErrors += numParseErrorsAlt
```

Comment:
I'd just overwrite the `numParseErrors` instead of adding them. It is only ever used for counters. And since the per TID and per PID maps should be identical, you are basically reporting doubled errors counter in this case.